Fix inventory cache interface (#50446)
ansible/ansible: docs/docsite/rst/dev_guide/developing_inventory.rst

The specifics will vary depending on API and structure returned. But one thing to keep in mind, if the inventory source or any other issue crops up you should ``raise AnsibleParserError`` to let Ansible know that the source was invalid or the process failed.

For examples on how to implement an inventory plugin, see the source code here:
`lib/ansible/plugins/inventory <https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/inventory>`_.

.. _inventory_plugin_caching:

inventory cache
^^^^^^^^^^^^^^^

> If you used the Cacheable base class and have extended the inventory plugin documentation with the inventory_cache documentation fragment you have caching at your disposal.

Reviewer: Is this the only way to get caching? If so, I'd recommend stating this as a direction rather than "if/then" (because "if/then" suggests there's an alternative - "if not/then"):

"To enable caching, use the ``Cacheable`` base class and extend the inventory plugin documentation with the ``inventory_cache`` documentation fragment."

> Now you can load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. Accordingly, you can skip this first step. Otherwise, for plugins that don't use ``_read_config_data``, the cache will need to be loaded with ``load_cache_plugin``.

Reviewer: ```suggestion
Next, load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. If your inventory plugin does not use ``_read_config_data``, you must load the cache explicitly with ``load_cache_plugin``.
```

.. code-block:: python

    NAME = 'myplugin'

    def parse(self, inventory, loader, path, cache=True):
        super(InventoryModule, self).parse(inventory, loader, path)

        self.load_cache_plugin()

> Before using the cache, retrieve a unique cache key using the ``get_cache_key`` method. This needs to be done by all inventory modules using the cache.

Reviewer: Information on why this needs to be done might help users here - for security? to enable two-way communication? Something like, "Before using the cache, you must retrieve a unique cache key . . . . method. This required step establishes a secure connection." (or whatever the actual reason).

.. code-block:: python

    def parse(self, inventory, loader, path, cache=True):
        super(InventoryModule, self).parse(inventory, loader, path)

        self.load_cache_plugin()
        cache_key = self.get_cache_key(path)

> The last parameter of ``parse`` is ``cache``. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ``self._cache`` like a dictionary. The following pattern allows refreshing the inventory to work in conjunction with caching.

Reviewer: This paragraph contains good info, and a better introductory sentence would make it great. Unfortunately I'm not sure what the main topic is, so I'm not sure what to suggest. It might be: "Now that you've enabled caching and loaded the correct plugin, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method." Or it might be:  "Be careful with the ``cache`` parameter, which controls data flow into and out of the cache; you want to avoid overwriting inventory at the wrong times." Again, detailing why this information is important helps users (and me!) understand it.

Author: I think I made this better. The topic is the flow of data/logic.

        user_cache_setting = self.get_option('cache')

        # read if the user has caching enabled and the cache isn't being refreshed
        attempt_to_read_cache = user_cache_setting and cache
        # update if the user has caching enabled and the cache is being refreshed; update this value to True if the cache has expired below
        cache_needs_update = user_cache_setting and not cache

        # attempt to read the cache if inventory isn't being refreshed and the user has caching enabled
        if attempt_to_read_cache:
            try:
                results = self._cache[cache_key]
            except KeyError:
                # This occurs if the cache_key is not in the cache or if the cache_key expired, so the cache needs to be updated
                cache_needs_update = True

        if cache_needs_updates:
            results = self.get_inventory()

            # set the cache
            self._cache[cache_key] = results

        self.populate(results)

After the ``parse`` method is complete, the contents of ``self._cache`` is used to set the cache plugin if the contents of the cache have changed.

> You also have cache methods ``set_cache_plugin``, ``update_cache_if_changed``, and ``clear_cache`` available. If you need to set the cache plugin before ``parse`` completes for some reason, you can use ``self.set_cache_plugin()`` which will force the cache plugin to be set with the contents of ``self._cache``. Alternatively, ``self.update_cache_if_changed()`` can be used to only set the cache plugin if ``self._cache`` has been modified. If you need to remove the keys in ``self._cache`` from your cache plugin, ``self.clear_cache()`` will delete those keys from the cache plugin.

Reviewer: Here I'd either fill in the "why" or make this a list, like this:
```
You have three other cache methods available:
- ``set_cache_plugin`` forces the cache plugin to be set with the contents of ``self._cache``
- ``update_cache_if_changed`` sets the cache plugin only if ``self._cache`` has been modified
- ``clear_cache`` deletes the keys in ``self._cache`` from your cache plugin
```

    extends_documentation_fragment:
      - inventory_cache

.. code-block:: python

    class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):

        NAME = 'myplugin'

Next, load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. If your inventory plugin does not use ``_read_config_data``, you must load the cache explicitly with ``load_cache_plugin``.

.. code-block:: python

    NAME = 'myplugin'

    def parse(self, inventory, loader, path, cache=True):
        super(InventoryModule, self).parse(inventory, loader, path)

        self.load_cache_plugin()

Before using the cache, retrieve a unique cache key using the ``get_cache_key`` method. This needs to be done by all inventory modules using the cache, so you don't use/overwrite other parts of the cache.

.. code-block:: python

    def parse(self, inventory, loader, path, cache=True):
        super(InventoryModule, self).parse(inventory, loader, path)

        self.load_cache_plugin()
        cache_key = self.get_cache_key(path)

> Now that you've enabled caching and loaded the correct plugin, you can set up the flow of data between the cache and your inventory using the `cache` parameter of the `parse` method. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ``self._cache`` like a dictionary. The following pattern allows refreshing the inventory to work in conjunction with caching.

Reviewer: ```suggestion
Now that you've enabled caching, loaded the correct plugin, and retrieved a unique cache key, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ``self._cache`` like a dictionary. The following pattern allows refreshing the inventory to work in conjunction with caching.
```
